src/blocked.c

```diff
@@ -351,265 +420,44 @@ void serveClientsBlockedOnSortedSetKey(robj *o, readyList *rl) {
-                    consumer = streamLookupConsumer(group,name,SLC_DEFAULT);
-                    if (consumer == NULL) {
-                        consumer = streamCreateConsumer(group,name,rl->key,
-                                                        rl->db->id,SCC_DEFAULT);
-                        if (noack) {
-                            streamPropagateConsumerCreation(receiver,rl->key,
-                                                            receiver->bpop.xread_group,
-                                                            consumer->name);
-                        }
-                    }
-                }
-
-                /* Emit the two elements sub-array consisting of
-                 * the name of the stream and the data we
-                 * extracted from it. Wrapped in a single-item
-                 * array, since we have just one key. */
-                if (receiver->resp == 2) {
-                    addReplyArrayLen(receiver,1);
-                    addReplyArrayLen(receiver,2);
-                } else {
-                    addReplyMapLen(receiver,1);
-                }
-                addReplyBulk(receiver,rl->key);
-
-                streamPropInfo pi = {
-                    rl->key,
-                    receiver->bpop.xread_group
-                };
-                streamReplyWithRange(receiver,s,&start,NULL,
-                                     receiver->bpop.xread_count,
-                                     0, group, consumer, noack, &pi);
-                /* Note that after we unblock the client, 'gt'
-                 * and other receiver->bpop stuff are no longer
-                 * valid, so we must do the setup above before
-                 * the unblockClient call. */
-
-unblock_receiver:
-                updateStatsOnUnblock(receiver, 0, elapsedUs(replyTimer), server.stat_total_error_replies != prev_error_replies);
-                unblockClient(receiver);
-                afterCommand(receiver);
-                server.current_client = old_client;
+            robj *o = lookupKeyReadWithFlags(rl->db, rl->key, LOOKUP_NONOTIFY | LOOKUP_NOSTATS);
+            /* 1. In case new key was added/touched we need to verify it satisfy the
+             *    blocked type, since we might process the wrong key type.
+             * 2. We want to serve clients blocked on module keys
+             *    regardless of the object type: we don't know what the
+             *    module is trying to accomplish right now.
+             * 3. In case of XREADGROUP call we will want to unblock on any change in object type
+             *    or in case the key was deleted, since the group is no longer valid. */
+            if ((o != NULL && (receiver->bstate.btype == getBlockedTypeByType(o->type))) ||
+                (o != NULL && (receiver->bstate.btype == BLOCKED_MODULE)) ||
+                ((receiver->bstate.btype == BLOCKED_STREAM) && strcasecmp(receiver->lastcmd->fullname, "xreadgroup") == 0))
```

Comment:
NOTE to self: fix this "ugly" xreadgroup part to consider rename of the command

Reply:
@ranshid sorry for the XREADGROUP special case, it was a bad idea

anyway, it seems like we're going in the direction of trying to unblock clients even if the key doesn't exist anymore (see https://github.com/redis/redis/pull/11310)

perhaps we want to keep this info inside the `bstate`? i.e. that we want to try to unlock if the key doesn't exist anymore (will be true for modules and XREADGROUP)
maybe we can even extend this idea and allow module to explicitly state if they want to be unblocked when the key doesn't exist? we can add a special ctx flag that the caller has to set before calling `RM_BlockClientOnKeys`
all that i mentioned above is an optimization, trying to aoid redundant calls to `unblockClientOnKey`/`moduleUnblockClientOnKey`

Reply:
then we can replace
```
|| ((receiver->bstate.btype == BLOCKED_STREAM) && strcasecmp(receiver->lastcmd->fullname, "xreadgroup") == 0)
```
with
```
|| (o == NULL && receiver->bstate.no_key_unblocks)
```

Reply:
@oranagra thoughts?

Reply:
I agree. Seems better that the command that initiated the blocking will tell us what's the trigger for unblocking, rather than have the blocking framework full of special code to be familiar with the different commands..

Reply:
> seems like my work in #11310 (signal deleted keys only if needed) and the work here (process the client on a deleted key only if needed) complete each other :)

Yes will be "fun" to merge these

> Q: why didn't you check `o == NULL`?
well, since this call will be triggered by both cases (type change and removed key) I need to attend both

Reply:
ok good point (is there a test that hits that? IIRC i wrote one)